### DISCUSSION TITLE
Arduino Mega 2560 — expand I/O with 4051 Analog Multiplexer?

### DISCUSSION THREAD

**Discussion by stllvdv** [shanteacontrols/OpenDeck#210] [Q&A] [answered]
Hey there,

Please correct me if necessary, I'm not sure I got that right:

a) the OpenDeck firmware not only runs on the official boards, but also supports something like the Arduino Mega 2560? b) I wonder if it's possible to expand the 16 analog inputs of an Arduino Mega 2560 with one or more 4051 analog multiplexer/demultiplexer units? Surely 8 additional inputs for each of the 16 analog inputs would create quite a few possibilities for many, many potentiometers.

Warm regards.

**Comment by paradajz** [comment, marked as answer]
> a) the OpenDeck firmware not only runs on the official boards, but also supports something like the Arduino Mega 2560?

Yes.

> b) I wonder if it's possible to expand the 16 analog inputs of an Arduino Mega 2560 with one or more 4051 analog multiplexer/demultiplexer units?

Yes. 

However, instead of Mega I really suggest using Raspberry Pi Pico because:

* it's cheaper
* it's smaller
* it's easier to flash
* it has loads of flash memory so it supports 10 presets
* the latency is better (on Mega everything goes through ATmega16u2 MCU which acts as an USB interface)

**Reply by stllvdv** [reply]
> b) I wonder if it's possible to expand the 16 analog inputs of an Arduino Mega 2560 with one or more 4051 analog multiplexer/demultiplexer units?

Circling back around that idea of expanding the analog inputs. Adding a couple of 4051 multiplexers to the 2560 shouldn't be difficult… would [OpenDeck Configurator](https://config.shanteacontrols.com/) recognize those additional analog inputs automatically?

**Reply by paradajz** [reply]
If you recompile the firmware and reflash it, then yes.

**Reply by stllvdv** [reply]
Understood. I'll look into that, but probably not yet fit/smart enough for that.

**Reply by paradajz** [reply]
https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant

**Reply by stllvdv** [reply]
Same would be true if I'd be using the Raspberry Pi Pico I guess? As the RP2040 also has only 3 analog inputs if I'm not mistaken? So would have to expand and custom-recompile that as well?

**Reply by stllvdv** [reply]
> Same would be true if I'd be using the Raspberry Pi Pico I guess?

https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant#mux-on-mux

**Reply by paradajz** [reply]
Yes. Check the [config/target](https://github.com/shanteacontrols/OpenDeck/tree/master/config/target) directory for examples.

**Reply by eurorackworld** [reply]
Using a pico is way easier in my opinion. Creating the config is easy and compiling it too (with a little help at first:P)), but flashing the firmware is of course a breeze on the pico due to just uploading the uf2 file onto the pico that shows up as a mounted drive on your pc.

**Reply by stllvdv** [reply]
@eurorackworld  Thanks, sounds great, I'll be sure to look into the RP2040 now! 

Doing all of this for the first time is more "complex" than I was hoping for, but let's see; never seen a breadboard, never used Ubuntu, never flashed a board before, but then yesterday I did just that with the Mega 2560 (with a little help of @paradajz) and I am now playing around with three potentiometers and a switch to bend cutoff/resonance/etc at will! Fun indeed!

**Reply by stllvdv** [reply]
So… any of the following might work to expand the [RP2040](https://www.reichelt.de/hr/de/raspberry-pi-pico-rp2040-cortex-m0-microusb-rasp-pi-pico-p295706.html?&trstct=vrt_pdn&nbc=1) with more GPIOs? Multiple [MCP 23017-E/SP](https://www.reichelt.de/hr/de/i-o-erweiterung-16bit-1-8-5v-seriell-i2c-dip-28-mcp-23017-e-sp-p140074.html) or [CD74HC4067](https://www.sparkfun.com/products/9056) or something like [Pico Decker](https://www.digikey.de/de/products/detail/pimoroni-ltd/PIM555/13981740) or [Quad Expander](https://www.berrybase.de/quad-gpio-expander-fuer-raspberry-pi-pico) etc…? Also, because I'm really new to this: adding new general purpose IOs means that those can be used for either digital or analog inputs?

**Reply by paradajz** [reply]
> Multiple MCP 23017-E/SP

No, not supported. Read the entire page to see what's supported: https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant. 4067 is supported yes.

3rd party extension boards aren't supported.

> adding new general purpose IOs means that those can be used for either digital or analog inputs?

No.

**Reply by stllvdv** [reply]
@eurorackworld Just seen [RPI Pico W + 4067 multiplexer](https://github.com/shanteacontrols/OpenDeck/discussions/177) where you commented as well. Was wondering if there's anything you could share in case I need help? Will probably try to add a few 4067 to the RP2040, looking for 48 (maybe 96) analog inputs (not sure how many digital inputs == switches/buttons I'll need).

**Reply by stllvdv** [reply]
Okay, slowly but surely getting into this, but some things are still confusing the heck out of me. I've now received a Pico, flashed it, works great; now what do I have to do to create a custom board variant with expanded analog input capabilities? Two goals: a) expand with a single 4051 for 8 inputs b) expand with 3x 4067 or more once I get the basics…

– get the packages, then clone the repository
– apparently update cmake as my version is always obsolete
– connect a MUX of choice **correctly** 
– configure the YAML **correctly** 
– build a new UF2 with `make TARGET=my_custom_conf` and flash

I kind of get how I have to connect the MUX in general, but the devil is in the details I guess. For starters: how do I know where on the PICO I have to connect `s1, s2, s3, (s4)` and `z0`? Which GPIO is which? How do I define these in the YAML? I've seen [the examples in the docs,](https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant#4067) but don't quite get it yet…

EDIT: Can I simply define those myself?

**Reply by stllvdv** [reply]
@paradajz Still exploring how to set up the config-files properly, but just ran `make TARGET=pico` in the meantime to get the hang of it all. This creates a folder in `build` with various files and subfolders. Final step would be renaming `application.elf.bin` to e.g. `pico.uf2` to be dropped to the PICO?

**Reply by stllvdv** [reply]
OK! Got the PICO and 1x 4067 working! Next goal/mystery: c) shift registers I guess to add… maybe 20… buttons!

**Comment by stllvdv** [comment]
Thanks! Love the project, brilliant! I'll be sure to look into that RP2040 as well then.

**Comment by paradajz** [comment]
> For starters: how do I know where on the PICO I have to connect `s1, s2, s3, (s4)` and `z0`? Which GPIO is which?

You pick any pin you want.

> Final step would be renaming application.elf.bin to e.g. pico.uf2 to be dropped to the PICO?

Not sure where you got this idea from? The build system creates uf2 file called `merged.uf2` in the build directory.

**Reply by stllvdv** [reply]
a) **Any** pin? Interesting… b) Doing all of this for the first time, might have misunderstood ["Build process creates binaries in build directory"](https://github.com/shanteacontrols/OpenDeck/wiki/Building-the-OpenDeck-firmware) and my mind connected binaries to .bin — completely overlooked that a .uf2 is created as well.

**Comment by stllvdv** [comment]


**Reply by paradajz** [reply]
Pull the latest master branch and try again, just run `make clean` first.

**Reply by stllvdv** [reply]
Thanks — `make clean` changes everything.

**Reply by paradajz** [reply]
It's mentioned in the wiki as well:

![image](https://github.com/shanteacontrols/OpenDeck/assets/2544094/fae6ffe8-9654-4c02-86a8-bd5ce5842e25)

https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant#building-custom-configuration
